### PR TITLE
Proposal: Convention for param/result reserved names

### PR DESCRIPTION
**Pull request by mattmoor**
This builds on https://github.com/tektoncd/community/pull/503 to actually define a convention for parameter and result names.

/hold

Putting a hold since this is dependent on the other PR (separate commits).

### CODE REVIEW

File: teps/0080-support-domainscoped-parameterresult-names.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+Allow Task and Pipeline authors to expose parameters and results that allow
+`.` characters.
+
+> _This TEP was originally raised [here](https://github.com/tektoncd/pipeline/issues/3590)
+as an issue for discussion._
+
+## Motivation
+
+The motivation of this work is to *enabl* us to establish conventions around the
```
Comment (vdemeester) [inline teps/0080-support-domainscoped-parameterresult-names.md:45]: `s/enabl/enable`

**Review by vdemeester — COMMENTED** [review]
Very naive question, what differs from #503 ? Is it just about "making" a *convention* out of what would results from #503 ?

File: teps/0081-convention-for-parameterresult-name-ownership-and-definition.md [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+Establish conventions for reserving segments of the parameter and result
+namespace, and enable namespace owners to define the types and semantics
+of well-known parameters and results.
+
+## Motivation
+
+### Goals
+
+The first goal of this TEP is to define a mechanism by which an organization
+(e.g. Tekton, Shipwright, Google, Redhat) may define parameters and results
```
Comment (chmouel) [inline teps/0081-convention-for-parameterresult-name-ownership-and-definition.md:46]: `Red Hat`

Comment (mattmoor) [reply teps/0081-convention-for-parameterresult-name-ownership-and-definition.md:46]: I'll learn which is "right" someday 😉

File: teps/0080-support-domainscoped-parameterresult-names.md (hunk 1 above)
Comment (mattmoor) [reply teps/0080-support-domainscoped-parameterresult-names.md]: Fixed in the baseline PR and rebased

**Comment by mattmoor** [comment]
> Is it just about "making" a convention out of what would results from #503 ?

Yes, `#503` is just about allowing `.`, this is about establishing a convention akin to the use of domains in K8s labels/annotations or reverse-domain for Java packages et al.

File: teps/0081-convention-for-parameterresult-name-ownership-and-definition.md [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+    ```
+
+3. Containerd plugins (copied from [stargz-snapshotter](https://github.com/containerd/stargz-snapshotter#quick-start-with-kubernetes))
+
+    ```toml
+    # Use stargz snapshotter through CRI
+    [plugins."io.containerd.grpc.v1.cri".containerd]
+      snapshotter = "stargz"
+      disable_snapshot_annotations = false
+    ```
```
Comment (mattmoor) [inline teps/0081-convention-for-parameterresult-name-ownership-and-definition.md:321]: Another example that just came up in slack is annotations on OCI images use reversed-domains: https://github.com/opencontainers/image-spec/blob/main/annotations.md#rules

**Comment by mattmoor** [comment]
/hold cancel

I have rebased this now that TEP-0080 has landed, and added the precedent of OCI annotations.  This should be RFAL.

File: teps/0081-convention-for-parameterresult-name-ownership-and-definition.md (hunk 3 above)
Comment (mattmoor) [reply teps/0081-convention-for-parameterresult-name-ownership-and-definition.md:321]: incorporated

**Comment by bobcatfish** [comment]
@mattmoor I'm not totally understanding what it would mean to accept this TEP - i.e. what would we actually DO? (e.g. publish some docs, maintain a list of reserved scopes?) if I missed this info in the PR plz just point me in the right direction

**Comment by mattmoor** [comment]
@bobcatfish Yeah, good question.

It could be docs, but in some sense I see this (and acknowledgement of this agreement by some of the core stewards) as the sort of core meta artifact.

Beyond the scope of this, I think the subsequent artifacts (owned by Tekton) will be the sorts of conventions Tekton adopts using this (under `dev.tekton.*`), and how those manifest in the various repos (for instance, I think Catalog will have a strong voice in shaping those!).

**Comment by bobcatfish** [comment]
> I think the subsequent artifacts (owned by Tekton) will be the sorts of conventions Tekton adopts using this (under dev.tekton.*)

Do you have any examples of what these artifacts might look like?

> (for instance, I think Catalog will have a strong voice in shaping those!).

If you have any ideas of specifically how this would impact the catalog (maybe an example of an existing Task that would benefit) I think that would be great to include

I'm not sure if this is the kind of thing you have in mind, but a couple important "schemas" I think Tekton should define and care about are: details when an image has been built and params that are needed when you clone a git repo (and probably outputs as well to reflect what was actually cloned). Is that the kind of thing you image using something like `dev.tekton` for? e.g. if you're producing an image artifact, (borrowing the object syntax from TEP-0075) something like:

```yaml
     results:
       - name: "dev.tekton.builtImage"
         type: object
         properties:
           url:
            type: string
           digest:
            type: string
```

(nit could we do "tekton.dev/buildImage" instead...)

**Comment by mattmoor** [comment]
> Do you have any examples of what these artifacts might look like?

I'd say a mix of TEPs, and (for those) docs so it is clear how folks integrate with conventions like these.


> I'm not sure if this is the kind of thing you have in mind

Yeah, images that are produced are exactly one of the cases I had in mind, and probably one of the first cases we should tackle because the pattern matching Chains has to do today is extremely janky 😅 

I totally agree that TEP-0075 would be *very* complementary here, and avoid an explosion of things that could be objects, but I suspect we will quickly yearn for SchemaRefs 😁 


> could we do "tekton.dev/buildImage" instead

It's certainly an option.  K8s does this, but has limits like single-slash.  One thing I think we should think about (it can be here or elsewhere) is how we deal with higher-arity things.  To be fair, the answer here may just be TEP-0075, most of my thinking on this aspect predates that.

One of the things I had in mind was that folks could use `dev.tekton.build-image` if a single image was produced, but we could also use `dev.tekton.build-image.<name>` to support multiple named outputs (maybe both if one is the "main" output).

**Comment by bobcatfish** [comment]
> I'd say a mix of TEPs, and (for those) docs so it is clear how folks integrate with conventions like these.

I'd still like to see something more concrete as a deliverable but maybe agreeing that we think this direction makes sense is enough for now?

> One of the things I had in mind was that folks could use dev.tekton.build-image if a single image was produced, but we could also use dev.tekton.build-image.<name> to support multiple named outputs (maybe both if one is the "main" output).

hm it'd be nice if we could find a way to not start encoding _even more_ structure into these reserved names - it feels like an array of `dev.tekton.build-image` objects is probably what we'd really want here?

Thinking about it more maybe `dev.tekton.build-image` isn't the way we'd want to go for chains - I think we'd want to have some kind of duck typing vs. requiring any specific name for the result, do you agree? e.g. if you produce an object with the properties `image-url`and `image-digest` we assume you've built an image 🤔 🤔 🤔 

I think I'd have an easier time wrapping my head around this TEP if we were trying to agree on structure for some specific examples, e.g. what interface Chains will be using an expecting - that's probably a separate TEP but I'm just still not totally understanding what we'd be agreeing to here with this TEP approved - without the TEP there's nothing to stop anyone from using this kind of naming scheme.

**Pull request closed without merging**